src/logging.c

```diff
@@ -99,50 +147,82 @@ rcutils_ret_t rcutils_logging_initialize_with_allocator(rcutils_allocator_t allo
+    } else if (retval == RCUTILS_GET_ENV_DEFAULT || retval == RCUTILS_GET_ENV_ZERO) {
+      g_output_stream = stderr;
+    } else if (retval == RCUTILS_GET_ENV_ONE) {
+      g_output_stream = stdout;
+    }
 
-    if (NULL == ret_str) {
-      if (strcmp(colorized_output, "1") == 0) {
-        g_colorized_output = RCUTILS_COLORIZED_OUTPUT_FORCE_ENABLE;
-      } else if (strcmp(colorized_output, "0") == 0) {
-        g_colorized_output = RCUTILS_COLORIZED_OUTPUT_FORCE_DISABLE;
-      } else if (strcmp(colorized_output, "") != 0) {
+    // Allow the user to choose how buffering on the stream works by setting
+    // RCUTILS_CONSOLE_BUFFERED_STREAM.  With an empty environment variable, use the
+    // default of the stream.  With a value of 0, force the stream to be unbuffered.
+    // With a value of 1, force the stream to be buffered.
+    retval =
+      rcutils_get_env_var_zero_or_one(
+      "RCUTILS_CONSOLE_BUFFERED_STREAM", "not buffered",
+      "buffered");
+    if (retval == RCUTILS_GET_ENV_ERROR) {
+      return RCUTILS_RET_INVALID_ARGUMENT;
+    } else if (retval == RCUTILS_GET_ENV_ZERO) {
+      if (setvbuf(g_output_stream, NULL, _IONBF, 0) != 0) {
         fprintf(
-          stderr,
-          "Warning: unexpected value [%s] specified for RCUTILS_COLORIZED_OUTPUT. "
-          "Output will be colorized if target stream is a terminal."
-          " Valid values are 0 and 1.\n",
-          colorized_output);
+          stderr, "Error setting stream buffering mode: %s\n", strerror(errno));
+        RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+          "Error setting stream buffering mode: %s", strerror(
+            errno));
+        return RCUTILS_RET_ERROR;
+      }
+    } else if (retval == RCUTILS_GET_ENV_ONE) {
+      if (setvbuf(g_output_stream, NULL, _IOLBF, 0) != 0) {
+        fprintf(
+          stderr, "Error setting stream buffering mode: %s\n", strerror(errno));
+        RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+          "Error setting stream buffering mode: %s", strerror(
+            errno));
+        return RCUTILS_RET_ERROR;
       }
-    } else {
-      RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
-        "Failed to get if output is colorized from env. variable [%s]. Using DEFAULT.",
-        ret_str);
-      ret = RCUTILS_RET_INVALID_ARGUMENT;
     }
+    // Note that for retval == RCUTILS_GET_ENV_DEFAULT, we do nothing to leave it at default.
+
+    retval = rcutils_get_env_var_zero_or_one(
+      "RCUTILS_COLORIZED_OUTPUT", "force color",
+      "force no color");
+    if (retval == RCUTILS_GET_ENV_ERROR) {
+      return RCUTILS_RET_INVALID_ARGUMENT;
+    } else if (retval == RCUTILS_GET_ENV_ZERO) {
+      g_colorized_output = RCUTILS_COLORIZED_OUTPUT_FORCE_DISABLE;
+    } else if (retval == RCUTILS_GET_ENV_ONE) {
+      g_colorized_output = RCUTILS_COLORIZED_OUTPUT_FORCE_ENABLE;
+    }
+    // Note that for retval == RCUTILS_GET_ENV_DEFAULT, we do nothing to leave it at default.
```

Comment:
@clalancette here and above, you may be able to better describe intent using a `switch` statement (i.e. instead of adding comments).